README.md
ICSC Library for Linux
======================

# WORK IN PROGRESS

This is an implementation of the ICSC serial communication protocol
for Linux. It is aimed at the Raspberry Pi and BeagleBone Black, but
other embedded Linux system will also work.

Building
--------

    $ autoreconf -fi
    $ ./configure
    $ make
    $ sudo make install

By default it installs into /usr/local/lib and /usr/local/include. To install it in
/usr instead, change the configure command:

    $ ./configure --prefix=/usr
    $ make
    $ sudo make install

It is possible to cross-compile the library on a Linux PC if you have the
Raspberry Pi toolchain installed:

    $ ./configure --host=arm-linux-gnueabihf --prefix=/path/to/place/to/put/it
    $ make
    $ make install

You can then SCP the files from /path/to/place/to/put/it to your Raspberry Pi.


debian/changelog
libicsc (1.0.0) unstable; urgency=low

  * Initial Release.

 -- Matt Jenkins <matt@majenko.co.uk>  Fri, 03 Jun 2016 16:18:00 +0000


src/gpio.c
#include <fcntl.h>
#include <sys/stat.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include "icsc.h"
#include "config.h"

static const char *gpio_root = "/sys/class/gpio";

int icsc_gpio_open(int num, int mode) {
    char temp[200];
    struct stat sb;
    int fd;

    // Check to see if the GPIO has been exported.
    snprintf(temp, 200, "%s/gpio%d", gpio_root, num);
    if (stat(temp, &sb) != 0) {
        // Not exported, so export it.
        snprintf(temp, 200, "%s/export", gpio_root);
        fd = open(temp, O_WRONLY);
        if (fd < 0) {
            icsc_error("Unable to export GPIO%d: %s\n", num, strerror(errno));
            return -1;
        }
        snprintf(temp, 200, "%d\n", num);
        write(fd, temp, strlen(temp));
        close(fd);
    }

    snprintf(temp, 200, "%s/gpio%d/direction", gpio_root, num);
    fd = open(temp, O_WRONLY);
    if (fd < 0) {
        icsc_error("Unable to export GPIO%d: %s\n", num, strerror(errno));
        return -1;
    }

    switch(mode) {
        case ICSC_GPIO_OUTPUT:
            write(fd, "out\n", 4);
            break;
        case ICSC_GPIO_INPUT:
            write(fd, "in\n", 3);
            break;
    }

    close(fd);

    snprintf(temp, 200, "%s/gpio%d/value", gpio_root, num);
    fd = open(temp, O_RDWR);
    if (fd < 0) {
        icsc_error("Unable to export GPIO%d: %s\n", num, strerror(errno));
        return -1;
    }
    close(fd);

    return num;
}

int icsc_gpio_read(int num) {
    char temp[200];
    char c;
    int fd;
    if (num < 0) {
        return;
    }
    snprintf(temp, 200, "%s/gpio%d/value", gpio_root, num);
    fd = open(temp, O_RDWR);
    if (fd < 0) {
        icsc_error("Unable to read GPIO%d: %s\n", num, strerror(errno));
        return -1;
    }
    read(fd, &c, 1);
    close(fd);

    if (c == '1') {
        return 1;
    }
    return 0;
}

int icsc_gpio_write(int num, int level) {
    char temp[200];
    int fd;
    if (num < 0) {
        return;
    }
    snprintf(temp, 200, "%s/gpio%d/value", gpio_root, num);
    fd = open(temp, O_RDWR);
    if (fd < 0) {
        icsc_error("Unable to write GPIO%d: %s\n", num, strerror(errno));
        return -1;
    }
    write(fd, level ? "1" : "0", 1);
    close(fd);
    return 0;
}

int icsc_gpio_close(int num) {
    char temp[200];
    int fd;
    if (num < 0) {
        return;
    }
    snprintf(temp, 200, "%s/unexport", gpio_root);
    fd = open(temp, O_WRONLY);
    if (fd < 0) {
        icsc_error("Unable to unexport GPIO%d: %s\n", num, strerror(errno));
        return -1;
    }
    snprintf(temp, 200, "%d\n", num);
    write(fd, temp, strlen(temp));
    close(fd);
    return 0;
}


src/icsc.c
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <endian.h>
#include <pthread.h>
#include <stdarg.h>

#include "icsc.h"
#include "config.h"

int doDebug = 0;

void icsc_enable_debug() { doDebug = 1; }
void icsc_disable_debug() { doDebug = 0; }

void icsc_debug(const char *f, ...) {
    va_list arg;
    if (doDebug == 1) {
        va_start(arg, f);
        char format[strlen(f) + 15];
        sprintf(format, "ICSC DEBUG: %s", f);
        vfprintf(stderr, format, arg);
        va_end(arg);
    }
}

void icsc_error(const char *f, ...) {
    va_list arg;
    va_start(arg, f);
    char format[strlen(f) + 15];
    sprintf(format, "ICSC ERROR: %s", f);
    vfprintf(stderr, format, arg);
    va_end(arg);
}

static void icsc_assert_de(icsc_ptr icsc) {
    if (icsc->dePin < 0) {
        return;
    }
    icsc_gpio_write(icsc->dePin, 1);
}

static void icsc_deassert_de(icsc_ptr icsc) {
    if (icsc->dePin < 0) {
        return;
    }
    icsc_gpio_write(icsc->dePin, 0);
}

static int icsc_send_raw(icsc_ptr icsc, uint8_t origin, unsigned char station, char command, uint8_t len, const char *data) {
    int i;
    uint8_t cs = 0;

    if (icsc->uartFD < 0) {
        return -1;
    }

    pthread_mutex_lock(&icsc->uartMutex);

    icsc_assert_de(icsc);

    for (i = 0; i < ICSC_SOH_START_COUNT; i++) {
        icsc_serial_write(icsc->uartFD, SOH);
    }

    icsc_serial_write(icsc->uartFD, station);
    cs += station;

    icsc_serial_write(icsc->uartFD, origin);
    cs += origin;

    icsc_serial_write(icsc->uartFD, command);
    cs += command;

    icsc_serial_write(icsc->uartFD, len);
    cs += len;

    icsc_serial_write(icsc->uartFD, STX);

    for (i = 0; i < len; i++) {
        icsc_serial_write(icsc->uartFD, data[i]);
        cs += data[i];
    }

    icsc_serial_write(icsc->uartFD, ETX);
    icsc_serial_write(icsc->uartFD, cs);
    icsc_serial_write(icsc->uartFD, EOT);

    icsc_serial_flush(icsc->uartFD);
    icsc_deassert_de(icsc);
    pthread_mutex_unlock(&icsc->uartMutex);
    return 0;
}

static int icsc_respond_to_ping(icsc_ptr icsc, uint8_t station, uint8_t len, const char *data) {
    return icsc_send_raw(icsc, icsc->station, station, ICSC_SYS_PONG, len, data);
}

static int icsc_process(icsc_ptr icsc, unsigned long timeout) {
    char inch;
    int i;
    command_ptr scan;

    if (icsc == NULL) {
        return -1;
    }

    if (icsc->uartFD < 0) {
        return -1;
    }

    if (icsc_serial_wait_available(icsc->uartFD, timeout) <= 0) {
        return 0;
    }

    while (icsc_serial_available(icsc->uartFD) > 0) {
        inch = icsc_serial_read(icsc->uartFD);

        icsc_debug("Received 0x%02x from FD %d in phase %d\n", inch, icsc->uartFD, icsc->recPhase);

        switch (icsc->recPhase) {
            case 0: // Looking for header
                memcpy(&(icsc->header[0]), &(icsc->header[1]), 5);
                icsc->header[5] = inch;
                if ((icsc->header[0] == SOH) && (icsc->header[5] == STX) && (icsc->header[1] != icsc->header[2])) {
                    icsc->recCalcCS = 0;
                    icsc->recStation = icsc->header[1];
                    icsc->recSender = icsc->header[2];
                    icsc->recCommand = icsc->header[3];
                    icsc->recLen = icsc->header[4];

                    icsc_debug("Found valid header from %d to %d\n", icsc->recSender, icsc->recStation);
                    pthread_mutex_lock(&icsc->uartMutex);

                    for (i = 1; i < 4; i++) {
                        icsc->recCalcCS += icsc->header[i];
                    }
                    icsc->recPhase = 1;
                    icsc->recPos = 0;

                    if ((icsc->recStation != icsc->station) &&
                        (icsc->recStation != ICSC_BROADCAST ||
                         icsc->recStation != ICSC_SYS_RELAY )) {
                        icsc_reset(icsc);
                        break;
                    }

                    icsc_debug("Packet is for me!\n");

                    if (icsc->recLen == 0) {
                        icsc_debug("No payload. Skipping to phase 2\n");
                        icsc->recPhase = 2;
                    } else {
                        icsc_debug("Payload length %d\n", icsc->recLen);
                        icsc->buffer = (char *)malloc(icsc->recLen);
                    }
                }
                break;

            case 1: // Receive data
                icsc->buffer[icsc->recPos++] = inch;
                icsc->recCalcCS += inch;
                if (icsc->recPos == icsc->recLen) {
                    icsc_debug("Finished receiving data\n");
                    icsc->recPhase = 2;
                }
                break;

            case 2: // Check for ETX
                if (inch == ETX) {
                    icsc_debug("Got ETX\n");
                    icsc->recPhase = 3;
                } else {
                    icsc_debug("Expecting ETX but got 0x%02x\n", inch);
                    icsc_reset(icsc);
                }
                break;

            case 3: // Grab the checksum
                icsc->recCS = inch;
                icsc->recPhase = 4;
                icsc_debug("Received checksum byte of 0x%02x\n", inch);
                break;

            case 4: // Check for ETX and check the checksum.
                pthread_mutex_unlock(&icsc->uartMutex);
                if (inch == EOT) {
                    icsc_debug("Got EOT\n");
                    if (icsc->recCS == icsc->recCalcCS) {
                        icsc_debug("Checksum is valid.\n");

                        switch (icsc->recCommand) {
                            case ICSC_SYS_PING:
                                icsc_debug("Responding to ping\n");
                                icsc_respond_to_ping(icsc, icsc->recSender, icsc->recLen, icsc->buffer);
                                break;
                        }

                        for (scan = icsc->commandList; scan; scan = scan->next) {
                            if ((scan->commandCode == icsc->recCommand || scan->commandCode == ICSC_CATCH_ALL) && scan->callback) {
                                icsc_debug("Executing callback for command %c\n", scan->commandCode);
                                scan->callback(icsc, icsc->recSender, icsc->recCommand, icsc->recLen, icsc->buffer);
                            }
                        }
                    }
                } 
                icsc_reset(icsc);
        }
            
    }

    return 0;
}

static void *icsc_read_thread(void *arg) {
    icsc_ptr icsc = (icsc_ptr)arg;

    icsc_debug("Read thread executing\n");

    icsc->readThreadRunning = 1;
    while (icsc->readThreadRunning == 1) {
        icsc_process(icsc, 100000); // 100ms timeout
    }

    icsc_debug("Read thread finishing\n");
}

int icsc_register_command(icsc_ptr icsc, char command, callbackFunction func) {
    command_ptr newcmd;
    command_ptr scan;

    newcmd = (command_ptr)calloc(1, sizeof(command_t));

    if (newcmd == NULL) {
        fprintf(stderr, "ICSC: Cannot allocate command: %s\n", strerror(errno));
        return -1;
    }

    newcmd->commandCode = command;
    newcmd->callback = func;
    newcmd->next = NULL;

    if (icsc->commandList == NULL) {
        icsc->commandList = newcmd;
        return 0;
    }

    for (scan = icsc->commandList; scan->next; scan = scan->next);
    scan->next = newcmd;

    icsc_debug("Registered new command for code '%c'\n", newcmd->commandCode);

    return 0;
}

int icsc_unregister_command(icsc_ptr icsc, char command) {
    command_ptr scan;
    command_ptr tmp;

    if (icsc->commandList == NULL) {
        return -1;
    }

    // Is it the head?
    if (icsc->commandList->commandCode == command) {
        tmp = icsc->commandList->next;
        free(icsc->commandList);
        icsc->commandList = tmp;
        return 0;
    }

    // Scan through for it.
    for (scan = icsc->commandList; scan->next; scan = scan->next) {
        if (scan->next->commandCode == command) {
            icsc_debug("Unregistered command for code '%c'\n", scan->next->commandCode);
            tmp = scan->next->next;
            free(scan->next);
            scan->next = tmp;
            return 0;
        }
    }

    return 0;
}

icsc_ptr icsc_init_de(const char *uart, unsigned long baud, uint8_t station, int de) {
    icsc_ptr newicsc;
    int rc;

    newicsc = (icsc_ptr)calloc(1, sizeof(icsc_t));
    
    if (newicsc == NULL) {
        fprintf(stderr, "ICSC: Cannot allocate new ICSC endpoint: %s\n", strerror(errno));
        return NULL;
    }

    icsc_debug("Endpoint allocated\n");

    // First try and open the UART.
    newicsc->uartFD = icsc_serial_open(uart, baud);
    if (newicsc->uartFD < 0) {
        free(newicsc);
        return NULL;
    }

    icsc_debug("UART %s Opened. FD: %d\n", uart, newicsc->uartFD);

    newicsc->station = station;
    newicsc->dePin = de;

    // If we have a GPIO pin specified then open it and set it to listen mode.
    if (newicsc->dePin >= 0) {
        rc = icsc_gpio_open(newicsc->dePin, ICSC_GPIO_OUTPUT);
        if (rc < 0) {
            icsc_serial_close(newicsc->uartFD);
            free(newicsc);
            return NULL;
        }
        rc = icsc_gpio_write(newicsc->dePin, 0);
        if (rc < 0) {
            icsc_serial_close(newicsc->uartFD);
            free(newicsc);
            return NULL;
        }
    }

    // Now start the reading thread. 

    icsc_debug("Starting read thread\n");

    pthread_mutex_init(&newicsc->uartMutex, NULL);

    pthread_attr_t attr;
    rc = pthread_attr_init(&attr);
    if (rc != 0) {
        fprintf(stderr, "ICSC: Cannot start read thread: %s\n", strerror(errno));
        icsc_serial_close(newicsc->uartFD);
        free(newicsc);
        return NULL;
    }

    rc = pthread_create(&newicsc->readThread, &attr, &icsc_read_thread, newicsc);
    if (rc != 0) {
        fprintf(stderr, "ICSC: Cannot start read thread: %s\n", strerror(errno));
        icsc_serial_close(newicsc->uartFD);
        free(newicsc);
        return NULL;
    }

    pthread_detach(newicsc->readThread);

    pthread_attr_destroy(&attr);

    icsc_debug("Read thread started OK\n");

    return newicsc;
}

icsc_ptr icsc_init(const char *uart, unsigned long baud, uint8_t station) {
    return icsc_init_de(uart, baud, station, -1);
}

int icsc_send_array(icsc_ptr icsc, uint8_t station, char command, uint8_t len, const char *data) {
    return icsc_send_raw(icsc, icsc->station, station, command, len, data);
}

int icsc_send_string(icsc_ptr icsc, uint8_t station, char command, const char *str) {
    return icsc_send_raw(icsc, icsc->station, station, command, (uint8_t)strlen(str), str);
}

// In ICSC a "long" is a signed 32-bit value. Must be little-endian.
int icsc_send_long(icsc_ptr icsc, uint8_t station, char command, int32_t data) {
    int32_t le = htole32(data);
    return icsc_send_raw(icsc, icsc->station, station, command, 4, (const char *)&le); 
}

// In ICSC an "int" is a signed 16-bit value. Must be little-endian.
int icsc_send_int(icsc_ptr icsc, uint8_t station, char command, int16_t data) {
    int16_t le = htole16(data);
    return icsc_send_raw(icsc, icsc->station, station, command, 2, (const char *)&le); 
}

int icsc_send_char(icsc_ptr icsc, uint8_t station, char command, int8_t data) {
    return icsc_send_raw(icsc, icsc->station, station, command, 1, (const char *)&data); 
}


int icsc_broadcast_array(icsc_ptr icsc, char command, uint8_t len, const char *data) {
    return icsc_send_raw(icsc, icsc->station, ICSC_BROADCAST, command, len, data);
}

int icsc_broadcast_string(icsc_ptr icsc, char command, const char *str) {
    return icsc_send_raw(icsc, icsc->station, ICSC_BROADCAST, command, (uint8_t)strlen(str), str);
}

// In ICSC a "long" is a signed 32-bit value. Must be little-endian.
int icsc_broadcast_long(icsc_ptr icsc, char command, int32_t data) {
    int32_t le = htole32(data);
    return icsc_send_raw(icsc, icsc->station, ICSC_BROADCAST, command, 4, (const char *)&le); 
}

// In ICSC an "int" is a signed 16-bit value. Must be little-endian.
int icsc_broadcast_int(icsc_ptr icsc, char command, int16_t data) {
    int16_t le = htole16(data);
    return icsc_send_raw(icsc, icsc->station, ICSC_BROADCAST, command, 2, (const char *)&le); 
}

int icsc_broadcast_char(icsc_ptr icsc, char command, int8_t data) {
    return icsc_send_raw(icsc, icsc->station, ICSC_BROADCAST, command, 1, (const char *)&data); 
}

int icsc_close(icsc_ptr icsc) {
    void *res;
    if (icsc == NULL) {
        return -1;
    }

    icsc_debug("Closing ICSC channel\n");

    icsc->readThreadRunning = 0;
    pthread_join(icsc->readThread, &res);

    icsc_debug("Read thread joined\n");

    if (icsc->commandList != NULL) {
        command_ptr scan;
        command_ptr tmp;

        scan = icsc->commandList;
        while (scan != NULL) {
            tmp = scan->next;
            free(scan);
            scan = tmp;
        }

    }

    free(icsc);
    icsc_debug("Memory freed up\n");
    return 0;
}

int icsc_reset(icsc_ptr icsc) {
    if (icsc->buffer != NULL) {
        free(icsc->buffer);
        icsc->buffer = NULL;
    }
    icsc->recPhase = 0;
    icsc->recPos = 0;
    icsc->recLen = 0;
    icsc->recCommand = 0;
    icsc->recCS = 0;
    icsc->recCalcCS = 0;
    pthread_mutex_unlock(&icsc->uartMutex);
    return 0;
}



src/icsc.h
/** @file icsc.h
 *  @brief C libray for ICSC communication
 */

/** \mainpage
 * \copyright
 * Copyright (c) 2016, Majenko Technologies
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 * 
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 * 
 * * Redistributions in binary form must reproduce the above copyright notice, this
 *   list of conditions and the following disclaimer in the documentation and/or
 *   other materials provided with the distribution.
 * 
 * * Neither the name of Majenko Technologies nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */


#ifndef _ICSC_H
#define _ICSC_H

#include <stdint.h>
#include <termios.h>
#include <sys/types.h>
#include <pthread.h>

#ifdef __cplusplus
extern "C" {
#endif


#define ICSC_BROADCAST  0x00
#define ICSC_CMD_SYS    0x1F
#define ICSC_SYS_PING   0x05
#define ICSC_SYS_PONG   0x06
#define ICSC_SYS_QSTAT  0x07
#define ICSC_SYS_RSTAT  0x08
#define ICSC_SYS_RELAY  0x09

//When this is used during registerCommand all message will pushed
//to the callback function

// Packet wrapping characters, defined in standard ASCII table
#define SOH 1
#define STX 2
#define ETX 3
#define EOT 4

//The number of SOH to start a message
//some device like Raspberry was missing the first SOH
//Increase or decrease the number to your needs
#define ICSC_SOH_START_COUNT 1

struct icsc_command;

typedef struct icsc_command command_t;
typedef struct icsc_command *command_ptr;

typedef struct {
    int uartFD;
    int dePin;
    command_ptr commandList;
    uint8_t station;

    char header[5];

    char *buffer;

    uint8_t recPhase;
    uint8_t recPos;
    uint8_t recCommand;
    uint8_t recLen;
    uint8_t recStation;
    uint8_t recSender;
    uint8_t recCS;
    uint8_t recCalcCS;

    pthread_t readThread;
    int readThreadRunning;
    pthread_mutex_t uartMutex;
} icsc_t, *icsc_ptr;

// Format of command callback functions
typedef void(*callbackFunction)(icsc_ptr, unsigned char, char, unsigned char, char *);

// Structure to store command code / function pairs as a linked list
struct icsc_command {
    char commandCode;
    callbackFunction callback;
    struct icsc_command *next;
};


/* gpio.c */

/** \defgroup gpio
 *  \brief Helper functions to make working with GPIO easier.
 * @{
 */

#define ICSC_GPIO_OUTPUT 0
#define ICSC_GPIO_INPUT 1

/*! \brief Open a GPIO device in either input or output mode. Exports the GPIO node.
 *  \param num GPIO number
 *  \param mode Either ICSC_GPIO_INPUT or ICSC_GPIO_OUTPUT
 *  \return 0 on success or -1 on error.
 */
extern int icsc_gpio_open(int num, int mode);

/*! \brief Read the value of a GPIO
 *  \param num GPIO number
 *  \return 1 if the GPIO reads high, 0 if it reads low, or -1 on an error.
 */
extern int icsc_gpio_read(int num);

/*! \brief Set a GPIO to high or low.
 *  \param num GPIO number
 *  \param level 1 for logic high or 0 for logic low
 *  \return 0 on success or -1 on error.
 */
extern int icsc_gpio_write(int num, int level);

/*! \brief Close a GPIO. Unexports the GPIO node.
 *  \param num GPIO number
 *  \return 0 on success or -1 on error.
 */
extern int icsc_gpio_close(int num);

/** @} */

/* serial.c */
/** \defgroup serial
 *  \brief Helper functions to make working with serial devices easier.
 * @{
 */

/*! \brief Open a serial device at a specific baud rate
 *  \param path Path to the serial device (e.g., /dev/ttyAMA0)
 *  \param baud Symbolic baud rate for the port in the form Bxxx (e.g., B115200)
 *  \return The file descriptor for the newly opened port or -1 on an error.
 */
extern int icsc_serial_open(const char *path, unsigned long baud);

/*! \brief Wait for serial data to arrive up until the timeout expires
 *  \param fd The file descriptor of the port opened by icsc_serial_open()
 *  \param timeout The maximum number of microseconds to wait for data to arrive
 *  \return 1 if data is available, 0 if it timed out, or -1 on an error.
 */
extern int icsc_serial_wait_available(int fd, unsigned long timeout);

/*! \brief Look to see if serial data is available
 *  \param fd The file descriptor of the port opened by icsc_serial_open()
 *  \return 1 if data is available, 0 if there is none, or -1 on an error.
 */
extern int icsc_serial_available(int fd);

/*! \brief Read a byte from the serial port
 *  \param fd The file descriptor of the port opened by icsc_serial_open()
 *  \return A byte from the serial buffer, or -1 if no bytes are available.
 */
extern int icsc_serial_read(int fd);
/*! \brief Wait until all data sent to the serial port has been delivered to the wire
 *  \param fd The file descriptor of the port opened by icsc_serial_open()
 *  \return nothing
 */
extern void icsc_serial_flush(int fd);

/*! \brief Write a byte to a serial port
 *  \param fd The file descriptor of the port opened by icsc_serial_open()
 *  \param c The byte to write to the port
 *  \return 0 if the byte was written, -1 on error.
 */
extern int icsc_serial_write(int fd, uint8_t c);

/*! \brief Close the serial port
 *  \param fd The file descriptor of the port opened by icsc_serial_open()
 *  \return nothing
 */
extern void icsc_serial_close(int fd);
/** @} */

/* icsc.c */

/** \defgroup callbacks
 *  \brief Functions for dealing with commands and callback functions
 * @{
 */
#define ICSC_CATCH_ALL    0xFF

/*! \brief Register a new command callback
 *  \param icsc Pointer to an icsc context created using icsc_init() or icsc_init_de()
 *  \param command The character to use for the command
 *  \param func The callback function to call when the command is received
 *  \return 0 if the command was registered successfully, otherwise -1 on an error.
 */
extern int icsc_register_command(icsc_ptr icsc, char command, callbackFunction func);

/*! \brief Unregister an old command character.
 *  \param icsc Pointer to an icsc context created using icsc_init() or icsc_init_de()
 *  \param command The command character to unregister
 *  \return 0 if the command was unregistered successfully, otherwise -1 on an error.
 */
extern int icsc_unregister_command(icsc_ptr icsc, char command);

/** @}*/

/** \defgroup Initialization
 *  \brief Functions used for initializing and finisging with an ICSC instance
 *  @{
 */

/*! \brief Create a new ICSC context, initialize the hardware, and start listening
 *         for messages.
 *  \param uart The path name of the UART device to communicate with (e.g., /dev/ttyAMA0)
 *  \param baud The baud rate symbolic name in the form Bxxxx (e.g., B115200)
 *  \param station The station number of this device
 *  \param de The GPIO number to use for the RS-485 DE pin.
 *  \return The pointer to the newly created context.
 */
extern icsc_ptr icsc_init_de(const char *uart, unsigned long baud, uint8_t station, int de);

/*! \brief Create a new ICSC context, initialize the hardware, and start listening
 *         for messages.
 *  \param uart The path name of the UART device to communicate with (e.g., /dev/ttyAMA0)
 *  \param baud The baud rate symbolic name in the form Bxxxx (e.g., B115200)
 *  \param station The station number of this device
 *  \return The pointer to the newly created context.
 */
extern icsc_ptr icsc_init(const char *uart, unsigned long baud, uint8_t station);

/*! \brief Close an ICSC instance freeing the memory. Terminates all communication.
 *  \param icsc Pointer to an icsc context created using icsc_init() or icsc_init_de()
 *  \return 0 on success or -1 on error.
 */
extern int icsc_close(icsc_ptr icsc);

/** @} */


/** \defgroup sending
 *  \brief Functions used for sending data to a remote station
 *  @{
 */

/*! \brief Send an array of data (or struct as if it were an array) to a remote station
 *  \param icsc Pointer to an icsc context created using icsc_init() or icsc_init_de()
 *  \param station Destination station to send to
 *  \param command Command character to trigger at the remote station
 *  \param len The length of the array or size of the struct
 *  \param data The data to send
 *  \return 0 on success, -1 on error.
 */
extern int icsc_send_array(icsc_ptr icsc, uint8_t station, char command, uint8_t len, const char *data);

/*! \brief Send a text string to a remote station
 *  \param icsc Pointer to an icsc context created using icsc_init() or icsc_init_de()
 *  \param station Destination station to send to
 *  \param command Command character to trigger at the remote station
 *  \param str The string to send
 *  \return 0 on success, -1 on error.
 */
extern int icsc_send_string(icsc_ptr icsc, uint8_t station, char command, const char *str);

/*! \brief Send a 32-bit integer to a remote station
 *  \param icsc Pointer to an icsc context created using icsc_init() or icsc_init_de()
 *  \param station Destination station to send to
 *  \param command Command character to trigger at the remote station
 *  \param data The integer to send
 *  \return 0 on success, -1 on error.
 */
extern int icsc_send_long(icsc_ptr icsc, uint8_t station, char command, int32_t data);

/*! \brief Send a 16-bit integer to a remote station
 *  \param icsc Pointer to an icsc context created using icsc_init() or icsc_init_de()
 *  \param station Destination station to send to
 *  \param command Command character to trigger at the remote station
 *  \param data The integer to send
 *  \return 0 on success, -1 on error.
 */
extern int icsc_send_int(icsc_ptr icsc, uint8_t station, char command, int16_t data);

/*! \brief Send an 8-bit integer to a remote station
 *  \param icsc Pointer to an icsc context created using icsc_init() or icsc_init_de()
 *  \param station Destination station to send to
 *  \param command Command character to trigger at the remote station
 *  \param data The integer to send
 *  \return 0 on success, -1 on error.
 */
extern int icsc_send_char(icsc_ptr icsc, uint8_t station, char command, int8_t data);

/** @} */



/** \defgroup broadcast
 *  \brief Functions used for broadcasting data to all remote stations
 *  @{
 */

/*! \brief Send an array of data (or struct as if it were an array) to all stations
 *  \param icsc Pointer to an icsc context created using icsc_init() or icsc_init_de()
 *  \param command Command character to trigger at the remote station
 *  \param len The length of the array or size of the struct
 *  \param data The data to send
 *  \return 0 on success, -1 on error.
 */
extern int icsc_broadcast_array(icsc_ptr icsc, char command, uint8_t len, const char *data);

/*! \brief Send a text string to all remote stations
 *  \param icsc Pointer to an icsc context created using icsc_init() or icsc_init_de()
 *  \param command Command character to trigger at the remote station
 *  \param str The string to send
 *  \return 0 on success, -1 on error.
 */
extern int icsc_broadcast_string(icsc_ptr icsc, char command, const char *str);

/*! \brief Send a 32-bit integer to all remote stations
 *  \param icsc Pointer to an icsc context created using icsc_init() or icsc_init_de()
 *  \param command Command character to trigger at the remote station
 *  \param data The integer to send
 *  \return 0 on success, -1 on error.
 */
extern int icsc_broadcast_long(icsc_ptr icsc, char command, int32_t data);

/*! \brief Send a 16-bit integer to all remote stations
 *  \param icsc Pointer to an icsc context created using icsc_init() or icsc_init_de()
 *  \param command Command character to trigger at the remote station
 *  \param data The integer to send
 *  \return 0 on success, -1 on error.
 */
extern int icsc_broadcast_int(icsc_ptr icsc, char command, int16_t data);

/*! \brief Send an 8-bit integer to all remote stations
 *  \param icsc Pointer to an icsc context created using icsc_init() or icsc_init_de()
 *  \param command Command character to trigger at the remote station
 *  \param data The integer to send
 *  \return 0 on success, -1 on error.
 */
extern int icsc_broadcast_char(icsc_ptr icsc, char command, int8_t data);
/** @} */


/** \defgroup debugging
 *  \brief Functions used for debugging and error reporting
 *  @{
 */

/*! \brief Enable debug messages
 *  
 *  * Caution - this gets very noisy
 *  
 *  \param none
 *  \return nothing
 */
extern void icsc_enable_debug();

/*! \brief Disable debug messages
 *  
 *  \param none
 *  \return nothing
 */
extern void icsc_disable_debug();

/*! \brief Display a debug message to stderr
 *
 *  The string "ICSC DEBUG: " is prepended to the message. Parameters
 *  and formatting follow the same rules as printf().
 *
 *  This function only output anything if debugging is first enabled with
 *  icsc_enable_debug().
 */
extern void icsc_debug(const char *fmt, ...);


/*! \brief Display an error message to stderr
 *
 *  The string "ICSC ERROR: " is prepended to the message. Parameters
 *  and formatting follow the same rules as printf().
 */
extern void icsc_error(const char *fmt, ...);
/** @} */

#ifdef __cplusplus
}
#endif

#endif
